### PR TITLE
[Work in progress] Legacy cache accessor implementation

### PR DESCRIPTION
**Pull request by oldalton**
Basic legacy cache implementation.
Things still missing:
- authority metadata
- tests

### CODE REVIEW

File: IdentityCore/src/oauth2/MSIDTokenRequest.h [hunk 1]
```diff
@@ -32,4 +32,20 @@
 @property (readonly) NSString *redirectUri;
 @property (readonly) NSString *clientId;
 
+- (instancetype)init NS_UNAVAILABLE;
```
Comment (antrix1989) [inline IdentityCore/src/oauth2/MSIDTokenRequest.h]: we should mark "-new" method as NS_UNAVAILABLE as well.

File: IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheFormat.m [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                          resource:nil
+                           context:context
+                             error:error];
+}
+
+- (NSArray<MSIDToken *> *)getAllSharedClientRTsWithParams:(MSIDRequestParameters *)parameters
+                                                  context:(id<MSIDRequestContext>)context
+                                                    error:(NSError **)error
+{
+    NSArray *legacyTokens = [_dataSource itemsWithKey:[MSIDTokenCacheKey keyForAllItems]
```
Comment (unpluggedk) [inline IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheFormat.m]: Would telemetry for look up here be good?

File: IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheFormat.m [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+          context:(id<MSIDRequestContext>)context
+            error:(NSError **)error
+{
+    [[MSIDTelemetry sharedInstance] startEvent:[context telemetryRequestId]
+                                     eventName:MSID_TELEMETRY_EVENT_TOKEN_CACHE_WRITE];
+    
+    MSIDTelemetryCacheEvent *event = [[MSIDTelemetryCacheEvent alloc] initWithName:MSID_TELEMETRY_EVENT_TOKEN_CACHE_WRITE
+                                                                           context:context];
+    
+    NSURL *newAthority = token.authority; // TODO: replace with an actual authority
```
Comment (unpluggedk) [inline IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheFormat.m]: nit: type on newAthority

File: IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+- (BOOL)saveTokensWithRequestParams:(MSIDRequestParameters *)requestParams
+                           response:(MSIDTokenResponse *)response
+                            context:(id<MSIDRequestContext>)context
+                              error:(NSError **)error
+{
+    MSIDAccount *account = [[MSIDAccount alloc] initWithTokenResponse:response];
+    
+    if (response.isMultiResource)
+    {
+        // Save ADAL access token item
```
Comment (unpluggedk) [inline IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m]: nit: not ADAL, primary perhaps?

File: IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                                        requestParams:requestParams
+                                              context:context
+                                                error:error];
+        
+        if (!result)
+        {
+            return NO;
+        }
+        
+        // Create ADAL refresh token item
```
Comment (unpluggedk) [inline IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m]: same as above

File: IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    MSIDAccount *account = [[MSIDAccount alloc] initWithTokenResponse:response];
+    
+    if (response.isMultiResource)
+    {
+        // Save ADAL access token item
+        MSIDToken *accessToken = [[MSIDToken alloc] initWithTokenResponse:response
+                                                                  request:requestParams
+                                                                tokenType:MSIDTokenTypeAccessToken];
+        
+        BOOL result = [_primaryFormat saveAccessToken:accessToken
```
Comment (unpluggedk) [inline IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m]: _primaryFormat saveAccessToken needed if we already have primary in all Formats?

File: IdentityCore/src/cache/MSIDToken.h [hunk 7]
```diff
@@ -57,7 +57,7 @@ typedef NS_ENUM(NSInteger, MSIDTokenType)
 - (BOOL)isExpired;
 
 - (instancetype)initWithTokenResponse:(MSIDTokenResponse *)response
-                              request:(MSIDTokenRequest *)request
+                              request:(MSIDRequestParameters *)requestParams
                             tokenType:(MSIDTokenType)tokenType;
```
Comment (unpluggedk) [inline IdentityCore/src/cache/MSIDToken.h]: do we need token type here? 
can we do something like the following to determine tokenType from response?

    if (tokenResponse.idToken && tokenResponse.accessToken) {
        // access token
    }
    
    if (tokenResponse.idToken && tokenResponse.refreshToken) {
        // refreshToken token
    }
    
    if (!tokenResponse.idToken) {
        // adfs token
    }

**Comment by unpluggedk** [comment]
Just add telemetry where you see fit :P

File: IdentityCore/src/cache/MSIDToken.h (hunk 7 above)
Comment (oldalton) [reply IdentityCore/src/cache/MSIDToken.h]: Talked offline that token response contains both access token and refresh token, so it won't be possible to check that way. Also, it's good to be explicit about the token type.

**Pull request merged**